### PR TITLE
support for dynamic batch sizes at runtime

### PR DESCRIPTION
**Pull request by TheMenko**
This PR is a follow up to #485

### CODE REVIEW

File: src/app.rs [hunk 1]
```diff
@@ -129,6 +129,11 @@ impl App {
         let identity_manager = Arc::new(identity_manager);
         let database = Arc::new(db);
 
+        let provers = database.get_provers().await?;
+
+        // restore prover state from previus shutdown
```
Comment (gswirski) [inline src/app.rs:134]: nit: `previous`

File: src/database/mod.rs [hunk 2]
```diff
@@ -325,6 +327,74 @@ impl Database {
...
+        let mut tx = self.pool.begin().await?;
+
+        let query = sqlx::query(
+            r#"
+              DELETE FROM provers WHERE batch_size = $1  
+            "#,
+        )
+        .bind(batch_size as i64);
+
+        tx.execute(query).await?;
```
Comment (gswirski) [inline src/database/mod.rs:393]: nit: do we need an explicit transaction here? wouldn't `self.pool.execute(query)` work the same?

**Review by gswirski — CHANGES_REQUESTED** [review]
We should think about the migration plan. With this PR, there are two ways to initialize provers: from the ENV var, and from the database. There should be only one source of truth - the database.

Since we don't want to restart our production service unnecessarily (currently it takes minutes), we need to be careful how we proceed with the deployment of this. My suggestion would be:
* Keep the ENV var setting for now. When initializing the identiy_manager, we try to `restore_provers` from the db first. If and only if we get 0 provers, we fetch them from the ENV var (and insert into the db).
* Once all environments have been properly migrated (i.e. they have provers in the database), we remove the ENV var altogether
* To make future deployments and local setup easier, /insertIdentity endpoint should warn that there are no provers available and hint at addProver admin endpoint.

File: src/database/mod.rs [hunk 3]
```diff
@@ -325,6 +327,69 @@ impl Database {
...
+                prover::Prover {
+                    url,
+                    batch_size,
+                    timeout_s,
+                }
+            })
+            .collect::<prover::Provers>())
+    }
+
+    pub async fn write_prover(
```
Comment (gswirski) [inline src/database/mod.rs:356]: nit: `insert_prover` or even `insert_prover_configuration` to be more consistent with other db methods

File: src/database/prover.rs [hunk 4]
```diff
@@ -0,0 +1,7 @@
+pub type Provers = Vec<Prover>;
+
+pub struct Prover {
```
Comment (gswirski) [inline src/database/prover.rs:3]: nit: maybe `ProverConfiguration` for clarity?

**Review by gswirski — CHANGES_REQUESTED** [review]
Tagging @cichaczem to confirm that it's OK to allow only one URL per batch size (given he's experimenting with more prover replicas).

I think this PR still doesn't quite work. We have a set of existing provers in the ENV var. We deploy this change, restart the process... and now we still only have those provers in the ENV var. We can't use the API to create those provers in the database because we'll end up adding duplicate batch_sizes to the identity manager.

IMHO we should do it such that if we read prover configuration from the ENV var, we migrate it to the db right away. The next time sequencer starts, the ENV is no longer used. Does it makes sense to you, @cichaczem ? I'm assuming the next PR would completely remove dependence on ENV var once all sequencer instances have been updated with this.

**Comment by cichaczem** [comment]
Regarding `only one URL per batch size`. We should deploy additional instances using the infra setup if we want multiple provers for the same batch size. K8s can manage deployment and load balancing for us using 1-2 settings in the deployment description. In that manner, we should have only one URL per batch size.

**Comment by cichaczem** [comment]
Ditto on @gswirski comment about populating DB with the list of provers in the ENV. Let's deploy that first, make sure we have every prover in, and remove the ENV in a separate PR.

**Review by gswirski — APPROVED** [review]

**Pull request merged**